### DISCUSSION TITLE
Generate Files on build-time

### DISCUSSION THREAD

**Discussion by lucemans** [withastro/roadmap#96] [Proposal]
![image](https://user-images.githubusercontent.com/10339043/152044998-82130241-e5f9-4a7d-89d1-a79a8ef87b26.png)

`[...page].astro`
```ts
export async function getStaticPaths({rss, writeFile}) {
  const allPosts = Astro.fetchContent('../../../content/*.md');

  allPosts.forEach(post => {
    writeFile('assets/posts/' + post.slug + '.svg', generateBannerSVG(post.title, post.description, post.slug));
  });
}
```
`generateBannerSVG` returns `string`,

Final file structure
```
dist/
├── posts/
│   ├── my-first-post.html
└── assets/
    └── posts/
        └── my-first-post.svg
```

**Comment by lucemans** [comment]
`Astro.assets`
`Astro.addAsset`
`Astro.includeFile`
@FredKSchott

**Comment by FredKSchott** [comment]
writeFile sounds a lot like `fs.writeFile()`, which gets at the problems that Fuzzy has run into where you need to save files to `public` as source, which needs to be committed and managed in source code which isn't ideal. 

What if instead it was a special `writeAsset()` function that saved it into the `dist/assets/` folder in build, and in dev in saved it to memory (or some invisible cache in dev, like `.astro/assets`) that didn't need to be checked into git? That would also support @afuzzybear and what he's looking for in the font component.

That could also possibly support the Image component? More prototyping needed!

**Reply by aFuzzyBear** [reply]
>where you need to save files to public as source, which needs to be committed and managed in source code 

I do lose sleep thinking about this sometimes, sad as it might sound but for the purposes of component creation and truly flexing with Astro, this one wee issue seems to creep up at times, "where do I save this new file Im making" 

Right now my go to is to stick it in the the `public` this is where 'static'-non interfering files are places and copied 1:1 in build, so it is ideally places to receive files that are generated during 'build/run' time, one further step I place is to have `pre:build` scripts to generate the assets prior to the build script being actioned. These would be for generating search indexes, (got a working Stork search component) creating translation files for use with 'i18next', being a few examples. 

Having Astro component files generate files on the fly has I feel always been allowed with the inclusion of the `node:builtins` so this practice can be employed having a wrapper around `fs.writeFile()` etc would be nice however it shouldnt be introduced at the cost of removing the more native libraries that are available. I feel that this can be best addressed through better documentation on practices and standards not implementing an API for a practice that should be common to most* node developers and be taught to new developers as transferable skills. 

Ive never been more liberated when it comes to creating components than I am when using Astro. The collocation helps significantly, but I have felt that there is a second untapped potential when it comes to Astro components, slightly tangential but imagine the following:
```
---
// You have your server/ code that is ran only at buildtime, 
// This can be a data fetching, then converting and saving the file to disk/memory
// Then you return an export,

Astro.export = result() | new FunkyMethod | .......
----
<!-- And there is no HTML being returned. Turning the Astro component into an Astro Server Module...  -->

---
```

That was just a thought that I have been tossing about, but I do elude to an issue beyond the other workable issues that this bring and that is the million dollar astro question.

Where do we save our stuff?

Ideally having a limited set of options that best address certain use-cases, such as:

- Memory: This is where image transformations would occur, it make sense to have the main image file as source on the disk located within the `public/` from here any transformations would be handled by plugins assigning that to memory, then at build-time it would be able to know what image files to make and have a place that astro knows where to save them, and how to resolve their URLs etc. (This last step needs to be abstracted away as much as possible from the end-user, I feel that it should be a hand-over from dev to Astro, this would save a lot of potential issues from arising of many practices all being wrong,)

- Disk: Well where on the disk, and what goes where more precisely, the newly proposed `assets` directory would be ideal, a single directory that Astro knows and can resolve everything quite nicely, (forgive the naivety) 

Lastly this conversation would need to include I feel discussions on a, again forgive the awful term 'Global Cache' of sorts, where transformations can be stored, helping to save on performance, (dont wish for the component to redo all that expensive work every time HMR gets itchy).

Thank you for listening,

**Comment by FredKSchott** [comment]
One risk worth calling out (related to our SSR-readiness discussions) is that this proposal would work well inside of getStaticPaths(), since that is called during the build. 

However, you don't have access to any props inside of getStaticPaths() or during an SSR build (because props happen during page render, which in SSR-world could happen on the server post-build). 

So there is a separate problem here for anyone who wants to call `writeAsset()` using props. This is what I'm trying to figure out with @matthewp in the #90 discussion!
